### PR TITLE
Add Dedicated Linux Remote Build

### LINKED ISSUE #1421: Add kudulite build for Linux Dedicated and Elastic Premium + Improvements

### PR DESCRIPTION
**Pull request by ankitkumarr**
Fixes #1421

### CODE REVIEW

File: src/Azure.Functions.Cli/Helpers/KuduLiteDeploymentHelpers.cs [hunk 1]
```diff
@@ -58,6 +58,34 @@ public static async Task<DeployStatus> WaitForServerSideBuild(HttpClient client,
...
+        {
+            // There is a tracked Locking issue in kudulite causing Race conditions, so we have to use this API
+            // to gather deployment progress.
+            ColoredConsole.Write("Remote build in progress, please wait");
+            while (true)
+            {
+                using (var request = new HttpRequestMessage(HttpMethod.Get, new Uri("/api/isdeploying", UriKind.Relative)))
+                {
+                    var response = await client.SendAsync(request);
+                    response.EnsureSuccessStatusCode();
```
Comment (ankitkumarr) [inline src/Azure.Functions.Cli/Helpers/KuduLiteDeploymentHelpers.cs]: I should add a retry policy here

File: src/Azure.Functions.Cli/Actions/AzureActions/PublishFunctionAppAction.cs [hunk 2]
```diff
@@ -547,7 +589,7 @@ public async Task PublishZipDeploy(Site functionApp, Func<Task<Stream>> zipFileF
             }, 2);
         }
 
-        public async Task<DeployStatus> PerformServerSideBuild(Site functionApp, Func<Task<Stream>> zipFileFactory)
+        public async Task<DeployStatus> PerformServerSideBuild(Site functionApp, Func<Task<Stream>> zipFileFactory, Func<HttpClient, Task<DeployStatus>> deploymentStatusPollTask, bool attachRestrictedToken = true)
```
Comment (Hazhzeng) [inline src/Azure.Functions.Cli/Actions/AzureActions/PublishFunctionAppAction.cs]: Ah, nice refactoring!

File: src/Azure.Functions.Cli/Helpers/KuduLiteDeploymentHelpers.cs [hunk 3]
```diff
@@ -58,6 +58,27 @@ public static async Task<DeployStatus> WaitForServerSideBuild(HttpClient client,
...
         }
 
+        public static async Task<DeployStatus> WaitForDedicatedBuildToComplete(HttpClient client, Site functionApp)
+        {
+            // There is a tracked Locking issue in kudulite causing Race conditions, so we have to use this API
+            // to gather deployment progress.
+            ColoredConsole.Write("Remote build in progress, please wait");
+            while (true)
+            {
+                var json = await InvokeRequest<IDictionary<string, bool>>(client, HttpMethod.Get, "/api/isdeploying");
```
Comment (Hazhzeng) [inline src/Azure.Functions.Cli/Helpers/KuduLiteDeploymentHelpers.cs]: The retry pattern is added in https://github.com/Azure/azure-functions-core-tools/pull/1436/commits/351e3509fd54b75a7496edfef0dcf61ba231965d#diff-5daf59331ae3b7d5664b9eb70713cf2bR128

File: src/Azure.Functions.Cli/Actions/AzureActions/PublishFunctionAppAction.cs [hunk 4]
```diff
@@ -281,14 +281,14 @@ private async Task PublishFunctionApp(Site functionApp, GitIgnoreParser ignorePa
             var functionAppRoot = ScriptHostHelpers.GetFunctionAppRootDirectory(Environment.CurrentDirectory);
 
             // For dedicated linux apps, we do not support run from package right now
-            var isFunctionAppDedicated = !functionApp.IsDynamic && !functionApp.IsElasticPremium;
-            if (functionApp.IsLinux && isFunctionAppDedicated && RunFromPackageDeploy)
+            var isFunctionAppDedicatedLinux = functionApp.IsLinux && !functionApp.IsDynamic && !functionApp.IsElasticPremium;
```
Comment (Hazhzeng) [inline src/Azure.Functions.Cli/Actions/AzureActions/PublishFunctionAppAction.cs]: May wonder if we want to enable elastic premium 💳

**Comment by ankitkumarr** [comment]
I also need to add logic to remove those "build" app settings if they are present and we are not doing a remote build.

**Comment by Hazhzeng** [comment]
Tested on consumption! Works fine!

File: src/Azure.Functions.Cli/Common/Constants.cs [hunk 5]
```diff
@@ -78,6 +79,14 @@ public static class KuduLiteDeploymentConstants
         {
             public const int ArmTokenExpiryMinutes = 4;
             public const int StatusRefreshSeconds = 3;
+
+            public static readonly IDictionary<string, string> LinuxDedicatedBuildSettings = new Dictionary<string, string>
+            {
+                { "ENABLE_ORYX_BUILD", "true" },
+                { "SCM_DO_BUILD_DURING_DEPLOYMENT", "1" },
+                { "BUILD_FLAGS", "UseExpressBuild" },
+                { "XDG_CACHE_HOME", "/tmp/.cache" }
```
Comment (ankitkumarr) [inline src/Azure.Functions.Cli/Common/Constants.cs]: Had to make this point to a local dir to optimize build times on Elastic Premium plans. I will try and change it to `/var` instead after testing it a bit to make sure temp cleanups don't affect the builds.

**Pull request merged**